### PR TITLE
[2444] Provide only valid view descriptions to the renderer

### LINKED ISSUE #2444: Provide only valid view description to renderer

### PR DESCRIPTION
**Pull request by mcharfadi**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/2444

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [x] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: doc/adrs/111_provide_only_valid_view_description_to_renderer.adoc [hunk 1]
```diff
@@ -0,0 +1,37 @@
+= ADR-000 - Adopt Architecture Decision Records
```
Comment (sbegaudeau) [inline doc/adrs/111_provide_only_valid_view_description_to_renderer.adoc]: ADR-111

File: doc/adrs/111_provide_only_valid_view_description_to_renderer.adoc [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+== Context
+
+Some rendered components  can have an invalid description. For example a reference widget component without a valid `Reference Name Expression`.
```
Comment (sbegaudeau) [inline doc/adrs/111_provide_only_valid_view_description_to_renderer.adoc]: One sentence per line

File: doc/adrs/111_provide_only_valid_view_description_to_renderer.adoc [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+This can provoke unexpected results. We want to avoid converting view dsl elements to programmatic elements if they have an invalid description.
+
+== Decision
+
+Only the `sirius-components-view-emf` package will be impacted.
+
+The class `ViewFormDescriptionConverterSwitch.java` will extends extends `FormSwitch<Optional<AbstractControlDescription>>` instead of `FormSwitch<AbstractControlDescription>`.
+
+`IWidgetConverterProvider#getWidgetConverter` will change to :
```
Comment (sbegaudeau) [inline doc/adrs/111_provide_only_valid_view_description_to_renderer.adoc]: No space before `:`

**Review by sbegaudeau — COMMENTED** [review]
The ADR also needs to be mentioned in the CHANGELOG.

File: doc/adrs/111_provide_only_valid_view_description_to_renderer.adoc (hunk 1 above)
Comment (mcharfadi) [reply doc/adrs/111_provide_only_valid_view_description_to_renderer.adoc]: done

File: doc/adrs/111_provide_only_valid_view_description_to_renderer.adoc (hunk 2 above)
Comment (mcharfadi) [reply doc/adrs/111_provide_only_valid_view_description_to_renderer.adoc]: done

File: doc/adrs/111_provide_only_valid_view_description_to_renderer.adoc (hunk 3 above)
Comment (mcharfadi) [reply doc/adrs/111_provide_only_valid_view_description_to_renderer.adoc]: done

**Review by sbegaudeau — DISMISSED** [review]

**Comment by pcdavid** [comment]
Looks fine to me, but there are other widgets which are created in an invalid state, notably the Pie Chart and Bar Chart (they default to null keys and values expressions, which cause NPEs and prevent the whole FormDescription to be converted).

**Review by pcdavid — APPROVED** [review]

**Pull request merged**